content/advent-2019/safe-use-of-unsafe-pointer.md

```diff
@@ -0,0 +1,325 @@
+not worked with `unsafe` before, I recommend reading my
+[previous Gopher Academy Advent series blog](https://blog.gopheracademy.com/advent-2017/unsafe-pointer-and-system-calls/)
+on the topic.
+
+Extra care and caution must be taken whenever introducing `unsafe` to a code
+base, but these diagnostic tools can help you solve problems before they lead to
+a major bug or a possible security flaw in your application.
+
+## Compile-time verification with `go vet`
+
+For several years now, the `go vet` tool has had the ability to check for
+invalid conversions between the `unsafe.Pointer` and `uintptr` types. 
+
+Let's take a look at an example program. Suppose we would like to use pointer
+arithmetic to iterate over and print each element of an array:
+
+```go
+package main
+
+import (
+    "fmt"
+    "unsafe"
+)
+
+func main() {
+    // An array of contiguous uint32 values stored in memory.
+    arr := []uint32{1, 2, 3}
+
+    // The number of bytes each uint32 occupies: 4.
+    const size = unsafe.Sizeof(uint32(0))
+
+    // Take the initial memory address of the array and begin iteration.
+    p := uintptr(unsafe.Pointer(&arr[0]))
+    for i := 0; i < len(arr); i++ {
+        // Print the integer that resides at the current address and then
+        // increment the pointer to the next value in the array.
+        fmt.Printf("%d ", (*(*uint32)(unsafe.Pointer(p))))
+        p += size
+    }
+}
+```
+
+At first glance, this program appears to work as expected, and we can see each
+of the array's elements printed to our terminal:
+
+```
+$ go run main.go 
+1 2 3
+```
+
+However, there is a subtle flaw in this program. What does `go vet` have to say?
+
+```
+$ go vet .
+# github.com/mdlayher/example
+./main.go:20:33: possible misuse of unsafe.Pointer
+```
+
+In order to understand this error, we must consult [the rules](https://golang.org/pkg/unsafe/#Pointer)
+of the `unsafe.Pointer` type:
+
+> Converting a Pointer to a uintptr produces the memory address of the value
+> pointed at, as an integer. The usual use for such a uintptr is to print it.
+>
+> Conversion of a uintptr back to Pointer is not valid in general.
+>
+> A uintptr is an integer, not a reference. Converting a Pointer to a uintptr
+> creates an integer value with no pointer semantics. Even if a uintptr holds
+> the address of some object, the garbage collector will not update that
+> uintptr's value if the object moves, nor will that uintptr keep the object
+> from being reclaimed.
+
+We can isolate the offending portion of our program as follows:
+
+```go
+p := uintptr(unsafe.Pointer(&arr[0]))
+
+// What happens if there's a garbage collection here? It's possible that the
+// address (now stored as a uintptr integer) is no longer valid!
+//
+// Let's assume p no longer points at a uint32. Perhaps this memory is now being
+// used to store a user's authentication credentials or a TLS private key.
+//
+// We've introduced a potential security flaw in our application!
```

Comment:
Maybe give an example of a security flaw? How can an attacker use it?

Reply:
I think I have covered that with the previous sentence and would rather not get too in-depth about a topic that I am not an expert in.